docs: KLIP-38 for Variable Substitution

spena opened the pull request:
### Description 
Add KLIP-38 to support variable substitution in ksqlDB.

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

colinhicks reviewed (commented):
Very exciting @spena! I had some initial questions mostly to check my understanding in my first pass.

spena commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,288 @@
+
+### 3. Referencing substitution variables
+
+Variables will be referenced by wrapping the variable between `{}` characters (i.e. `{replicas}`). These characters are also used by
```
I might change the prefix/suffix for '${}' to avoid conflicting with JSON literal strings.

colinhicks replied:
+1 for `${}`.

spena replied:
Done

spena commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,288 @@
+
+### 6. Implementation
+
+For CLI, variable substitution will happen in the CLI side before making a ksqlDB request to the server. This allows less changes in ksqlDB code as we have  
+the `SET CLI` to enable/disable, and variables will be living in the CLI server memory, so we have quickly access to them.
```
@colinhicks do we need to support Cloud? If we don't support REST API then we can't support it. Only way to support REST and Cloud is to open a session in the server, send all variables to the server so they're stored in that temporal session, and replace variables in the server side. It's more involved, but we'll allow Cloud users to use it.

colinhicks replied:
We can think of this as a feature of clients, where each client that supports string substitution applies any transformations before sending the SQL to the server.

The Confluent Cloud UI is a ksqlDB client that can implement support for substitution at some point. It does not need to support it initially.

colinhicks commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,288 @@
+
+### 4. Context for substitution variables
+
+Variable substitution will be allowed only in specific SQL statements, and only in places where literals or hard-coded values can be used.
+Variables cannot be used as column-names, reserved keywords, etc.
```
I think supporting identifiers like source name, column name, in addition to literals would be best.

I we should probably not support arbitrary statement fragments.

spena replied:
Done. I made a reference to support identifiers for column and source names.

colinhicks commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,288 @@
+Substitution will not attempt to add single-quotes to the values. Users will need to know the data type to use when using a variable. The ksqlDB
+server will be in charge of parsing and failing if the type is not allowed.
+
+Note: Variables are not case sensitive. A reference to `{replicas}` is the same as `{REPLICAS}`.
```
I think that makes sense. For reference MySQL's variables are also case-insensitive.

agavra commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,302 @@
+This looks confusing too. So we have to decide for a new syntax for substitution variables.
+
+Other DBs use different syntax to create user variables (DECLARE, DEFINE, \set, SET). For ksqlDB, I'll choose to use `DEFINE` and `UNDEFINE` which
+is what Oracle SQL*plus uses. Any user coming from this DB will immediately understand the command.  
```
can you link to the Oracle docs here for future reference (is it https://docs.oracle.com/cd/E18283_01/server.112/e16604/ch_twelve017.htm ?)

Though if I had to choose, I think the SQL+ syntax here is a little unintuitive. MySQL: https://docs.oracle.com/cd/E17952_01/mysql-5.7-en/user-variables.html uses something that to me is more intuitive (and is similar to SqlServer)

spena replied:
Oracle SQL*Plus
https://blogs.oracle.com/opal/sqlplus-101-substitution-variables

I wanted to use a similar character like Mysql. I also like the `@`, but the `StringSubstitutor` library doesn't allow me to us only the @. I have to add a suffix to the variable, so a `@var@` or `:var:` would work.

spena replied:
And regarding the `SET`. I didn't want to mix the current SET syntax which uses single-quotes for keynames, and I wasn't sure what other keyword to add to make reference to user-variables. Any proposal?

```
ksql> SET 'auto.offset.reset' = 'earliest';
ksql> SET VAR var1 = 'hola';
```

agavra replied:
If there's precedent (Hive, Impala, Spark and Presto) I think we can stick with your proposal as users will be accustomed to this already.

agavra commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,302 @@
+
+### 3. Referencing substitution variables
+
+Variables will be referenced by wrapping the variable between `${}` characters (i.e. `${replicas}`).
```
If we're following Oracle SQL+ why not use `&` or `&&` to reference the variables?

spena replied:
In Oracle, the `&` prompts users to type the variable value when not defined, and `&&` to reuse the variable after the user has typed a value. That behavior is out of scope, so we either use `&` or something else. And I had the problem with the `StringSubstitutor` that requires a suffix too, so `&var&`? 

Either we write our own string parsing or get in love with some characters to wrap the variables names :).

Btw, the `${var}` is used in Hive, Impala, Spark and Presto. We're not alone. They use `SET`, though, for setting variables, but they do not override server settings or they use a different name for user variables. For instance, Hive uses `SET hiveconf:var1 = 'value'` and it is referenced as `${hiveconf:var1}` just to differentiate between user variables and system overrides `${system:varSys1}`.

agavra commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,302 @@
+The error message is just an example of what ksqlDB will display.
+
+When using SQL scripts in headless mode, we need to provide a server setting to enable/disable variable substitution. The setting will be
+`ksql.enable.headless.variable.substitution` as Boolean (default = true). The reason is that `SET CLI` is a CLI-specific command not available
```
```suggestion
`ksql.headless.variable.substitution.enable` as Boolean (default = true). The reason is that `SET CLI` is a CLI-specific command not available
```

spena replied:
Done.

agavra commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,302 @@
+Variable substitution will be allowed in specific SQL statements. They can be used to replace text and non-text literals, and identifiers such as
+column names and stream/table names. Variables cannot be used as reserved keywords.
+
+For instance:
```
what if I don't want a literal `${foo}` to be replaced? e.g. I actually want `${foo}` to be the value in the statement (`SELECT '${foo}' as dolla_foo FROM bar`)? do i need to turn substitution off and then on after I issue the statement?

spena replied:
Yes,
```
ksql> set variable-substitution off;
ksql> SELECT '${foo}' as dolla_foo FROM bar;
ksql> set variable-substitution on;
```

I was playing with escape characters, but the `\${foo}` fails, and the `\\${foo}` results in `SELECT '\${foo}'`. 

I wanted to be fancier and behave like Mysql and Posgtres. They do not substitute variables inside quotes. For instance:
```
ksql> CREATE ... WITH (kafka_topic=@topicName);
or
ksql> CREATE ... WITH (kafka_topic=:topicName);
```
I would have to trick the `StringSubstitutor` a little bit to reject replacing variables inside quotes, it's not a problem. The problem I had was how to replace this:
```
ksql> define colName = 'name';
ksql> SELECT id, ${colName} FROM stream;
```
would become:
```
ksql> SELECT id, "name" FROM stream;   // valid statement 
or
ksql> SELECT id, name FROM stream;     // valid statement
```
Hard to know.  You can do prepared statements in Mysql, but that's very hacky.
https://stackoverflow.com/questions/4428761/mysql-field-name-from-variable

agavra replied:
I don't think implementation should drive the design. We should define what design we want here in the KLIP, and then if it's tough to implement we can issue a known bug report and fix it later.

cc @colinhicks do you think we need to hammer this issue out before we approve the KLIP?

colinhicks replied:
```
ksql> set variable-substitution off;
ksql> SELECT '${foo}' as dolla_foo FROM bar;
ksql> set variable-substitution on;
```

This seems like a reasonable limitation to me.

agavra reviewed (commented):
This LGTM! Some questions in line before green check

spena commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,295 @@
+# KLIP 34 - Variable Substitution
```
Done

vcrfxia commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,302 @@
+may be defined using a client method instead of a new syntax. This is to ensure users know the scope starts after the `create()` call.
+
+TBD: Define a Java API method for variable substitution.
+
```
Do we have plans to support variable substitution in the UI? The UI handles setting properties differently from the CLI: rather than the SET/UNSET syntax for session properties supported by the CLI, query properties instead have to be set using the dropdown in the UI editor. This divergence from the CLI has confused users in the past and I worry that having a different mechanism for defining variables in the UI will lead to similar confusion.

spena replied:
Yes, it's gonna be different in the UI than the CLI. I raised this question to @colinhicks too, and seems we'll have this feature separately from the UI. The reason? We'll have to work on opening temporary sessions between the client/server connections, then send variables to the server and stored them temporary there, then do the replacement in the server. That's more involved work.

We might need to do these changes in the future to make better UX for users.

vcrfxia commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,302 @@
+In Java API client, the session is opened when `Client.create()` is called, and closed when `Client.close()` is called. Variables
+may be defined using a client method instead of a new syntax. This is to ensure users know the scope starts after the `create()` call.
+
+TBD: Define a Java API method for variable substitution.
```
+1 to defining a new method for this rather than trying to piggyback off existing methods. If we don't support variable substitution in the REST API, the Java client will have to maintain variables in its own code, similar to the CLI. If we updated the CLI to use the Java client (https://github.com/confluentinc/ksql/issues/6269) we'd have fewer issues like these in the future but that's out of scope for this work for sure.

vcrfxia commented on design-proposals/klip-38-variable-substitution.md:
````diff
@@ -0,0 +1,302 @@
+SOURCE '<script_file>';
+```
+
+The `SOURCE` command will work like a `RUN SCRIPT`. If other SQL statements are in the file, then those statements will also be
````
What's the benefit in having `SOURCE` allow users to execute statements similar to `RUN SCRIPT`? It seems confusing to have two ways to do this with only a very subtle distinction (variables set in `SOURCE` persist into the sessions outside the scope of the command while those for `RUN SCRIPT` don't). Can we have `SOURCE` _only_ load variables from a file, and then ask users to use `RUN SCRIPT` afterwards if they want to also load statements?

spena replied:
We could that too. I'm not sure how users will use their scripts, though. Are they willing to split their scripts between vars and commands?

I just though of using the bash behavior:
```
$ cat /tmp/s1
export ID=1
echo id=$ID

$ cat /tmp/s2
export ID=2
echo id=$ID

$ source /tmp/s1
id=1
$ bash /tmp/s2
id=2

// Uses the ID set in by the source command
$ echo $ID
1
```

Mysql uses SOURCE too, and it allows variables and commands.
https://dev.mysql.com/doc/refman/5.7/en/mysql-batch-commands.html

agavra replied:
Does `MySQL` also have a `RUN SCRIPT`? I tend to agree with Victoria that this seems a little confusing.

vcrfxia commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,302 @@
+### 4. Context for substitution variables
+
+Variable substitution will be allowed in specific SQL statements. They can be used to replace text and non-text literals, and identifiers such as
+column names and stream/table names. Variables cannot be used as reserved keywords.
```
Is it normal for variables to not be allowed as keywords? I don't mind but if we go with this we need to make sure we have good error messages since a natural extension (unless it's just my naivete speaking?) from the example below:
```
ksql> CREATE STREAM ${streamName} (${colName1} INT, ${colName2} STRING) \
      WITH (kafka_topic='${topicName}', format='${format}', replicas=${replicas}, ...);
```
is to wonder about
```
ksql> CREATE STREAM ${streamName} (${colName1} ${colType1}, ${colName2} ${colType2}) \
      WITH (kafka_topic='${topicName}', format='${format}', replicas=${replicas}, ...);
```
which won't be accepted based on this proposal. Hopefully the error message makes it very clear why.

spena replied:
That really depends. Some big-data DBs allow any substitution, but other DBs (mysql, oracle, ...) do not allow replacing keywords. They have a different way to do that, though. 

Here's the discussion with @colinhicks about it.
https://github.com/confluentinc/ksql/pull/6259#discussion_r491181382

agavra replied:
I think given the goal (running these scripts in different environments) it's probably OK to not allow the types

vcrfxia replied:
Agreed it's fine to not allow keyword substitutions -- as long as we have good error messages. The error messages on syntax errors we have today are pretty abysmal... :)

vcrfxia commented on design-proposals/klip-38-variable-substitution.md:
```diff
@@ -0,0 +1,302 @@
+      WITH (kafka_topic='${topicName}', format='${format}', replicas=${replicas}, ...);
+      
+ksql> INSERT INTO ${streamName} (${colName1}, ${colName2}) \
+      VALUES (${val1}, '${val2}');
```
It's interesting to me that the contents of variables are substituted into the statement directly, and it's left to the user to get the quoting of the variables correct. This makes sense but it'd be great to have helpful error messages when this goes wrong. For example, if the user mistakenly submits
```
ksql> CREATE STREAM ${streamName} (${colName1} INT, ${colName2} STRING) \
      WITH (kafka_topic=${topicName}, format='${format}', replicas=${replicas}, ...);
```
instead of
```
ksql> CREATE STREAM ${streamName} (${colName1} INT, ${colName2} STRING) \
      WITH (kafka_topic='${topicName}', format='${format}', replicas=${replicas}, ...);
```
where the difference is that they're missing quotes around the topic name, it'd be great to have a useful error message so they aren't left guessing in the dark.

spena replied:
Right. I'll try to add a customized error. I was thinking on leverage the parsing errors, which displays an error when using `${topicName}`. I'll see if the second parser can detect those undesired places.

vcrfxia commented on design-proposals/klip-38-variable-substitution.md:
````diff
@@ -0,0 +1,302 @@
+ksql> SELECT * FROM ${streamName} \
+      WHERE ${colName1} == ${val1} and ${colName2} == '${val2}' EMIT CHANGES; 
+```
+Any attempt of using variables on non-permitted places will fail with the current SQL parsing error found when parsing the variable string.
````
> fail with the current SQL parsing error found when parsing the variable string

As mentioned above, I think we'll want better error messages here. The SQL parsing errors we see today are awful in terms of debuggability and I worry that variable substitution will make this all the more confusing. Can we make improving these error messages in scope for this work, or is that too much?

vcrfxia reviewed (commented):
Cool stuff @spena ! Comments/questions inline.

agavra reviewed (approved):
LGTM! This is an exciting feature in the CI/CD realm

derekjn commented:
Overall I think this is a good approach here. The only thing I'd like to raise is that I think that if we're using syntax for this, then substitution should happen server side. I think that processing syntax on the client could lead to confusion for users who are using HTTP or their own client implementation. But fundamentally, I feel that syntax should be completely independent of what client is being used.

spena commented:
@derekjn I was thinking on the scenario of users using raw HTTP requests. If we support variable substitution in the server-side, then they should work as the `SET` statements, which work only per request.
```
{ 
   statement: "
        DEFINE x = 'stream';

        SET 'auto.offset.reset' = 'earliest';
        SELECT * FROM ${x} EMIT CHANGES;
    "
}
```

This is for simple cases. Users can even read all statements from a file and make one single HTTP request. I don't know if that's something users are used to do, though. I would make one HTTP request per line if it was me.

Anyway, we should still support variable substitution in the CLI side. The CLI sends one request at a time. Otherwise, we'll have to send all the current variables on every request made by the user.

One advantage supporting server-side substitution might be when using the UI. I need to confirm it works like this, but I assume that the UI sends all statements in the UI textbox in one request. Like:
```
DEFINE x = 'stream';
SET 'auto.offset.reset' = 'earliest';
SELECT * FROM ${x} EMIT CHANGES;
```
If the above is sent as one request by the UI, then we'll have automatic support for the UI :). But need to confirm if that's the case, otherwise the UI will still have the same problem.

Btw, we need to add a configuration to override variable substitution in the HTTP request too. Like:
```
{ 
   statement: "
        DEFINE x = 'stream';

        SET 'ksql.variable.substitution.enable' = 'false'
        SELECT * FROM ${x} EMIT CHANGES;  // This fill fail 'cause ${x} is no substituted
    "
}
```

@colinhicks what do you think?

colinhicks commented:
> Anyway, we should still support variable substitution in the CLI side. The CLI sends one request at a time. Otherwise, we'll have to send all the current variables on every request made by the user.

Does this mean both both client-side substitution and server-side substitution apply to the CLI? Or would the CLI always use client-side substitution and not pass variables to the server?

spena commented:
@derekjn I added more details to how variable substitution will happen in the server-side. You were right about the new syntax introduction. We should allow variables in SQL scripts, and let users execute the script from any client they want, such as load the file in Java and send one HTTP request with all statements, or copy/paste a script file on the UI.

This is what I changed if you want to review it:

* Use of triple-quotes to escape single-quotes (see Creating variables)
* Add printing variables section only for CLI (see Printing variables)
* Added a context/scope for variable substitution in raw HTTP requests
* Replace `set cli variable-substutition` for `set 'ksql.variable.substitution.enable'` now that server-side substitution
  is the way to go.
* Edited the implementation section to explain how CLI will send requests for substitution

derekjn commented:
@spena LGTM 👍

spena commented:
Tests failure unrelated to the doc
